File: docs/current_docs/manuals/developer/python/snippets/services/service-lifecycle-3/main.py
from dagger import dag, function, object_type


@object_type
class MyModule:
    @function
    async def redis_service(self) -> str:
        # create Redis service container
        redis_srv = dag.container().from_("redis").with_exposed_port(6379).as_service()

        # create Redis client container
        redis_cli = (
            dag.container()
            .from_("redis")
            .with_service_binding("redis-srv", redis_srv)
            .with_entrypoint(["redis-cli", "-h", "redis-srv"])
        )

        # set and get value
        return (
            await redis_cli.with_exec(["set", "foo", "abc"])
            .with_exec(["get", "foo"])
            .stdout()
        )

File: docs/current_docs/manuals/developer/snippets/services/service-lifecycle-3/go/main.go
package main

import (
	"context"
)

type MyModule struct{}

// creates Redis service and client
func (m *MyModule) RedisService(ctx context.Context) (string, error) {
	redisSrv := dag.Container().
		From("redis").
		WithExposedPort(6379).
		AsService()

	// create Redis client container
	redisCLI := dag.Container().
		From("redis").
		WithServiceBinding("redis-srv", redisSrv)

    args := []string{"redis-cli", "-h", "redis-srv"}

	// set and get value
	return redisCLI.
		WithExec(append(args, "set", "foo", "abc")).
		WithExec(append(args, "get", "foo")).
		Stdout(ctx)
}
